remote/remote_test.go

```diff
@@ -501,6 +512,59 @@ func testImage(t *testing.T, when spec.G, it spec.S) {
 				})
 			})
 		})
+
+		when("#RetryLogic", func() {
+			var server *httptest.Server
+			testCases := []RemoteTestCase{
+				{
+					when:          "Manifest API in registry returns status code 200",
+					it:            "Do not retry after status code 200 ",
+					repo:          "org/do-not-retry",
+					statusCode:    http.StatusOK,
+					failedCount:   0,
+					expectedTries: 1,
+				},
+				{
+					when:          "Manifest API in registry returns status code 404",
+					it:            "Retry after status code 404",
+					repo:          "org/retry-not-found",
+					statusCode:    http.StatusNotFound,
+					failedCount:   2,
+					expectedTries: 3,
+				},
+				{
+					when:          "Manifest API in registry returns status code 401",
+					it:            "Retry after status code 401",
+					repo:          "org/retry-unauthorized",
+					statusCode:    http.StatusUnauthorized,
+					failedCount:   2,
+					expectedTries: 3,
+				},
+			}
+
+			for _, tc := range testCases {
+				tc := tc
+				when(tc.when, func() {
+					mockServer := h.NewMockServer(tc.repo, tc.statusCode, tc.failedCount)
+					it.Before(func() {
+						server = mockServer.Server()
+						u, err := url.Parse(server.URL)
+						h.AssertNil(t, err)
+						repoName = u.Hostname() + ":" + u.Port() + "/" + tc.repo
+					})
+
+					it(tc.it, func() {
+						defer server.Close()
+						_, err := remote.NewImage(repoName, authn.DefaultKeychain, remote.WithPreviousImage(repoName))
+						h.AssertNil(t, err)
+						if tc.expectedTries != mockServer.ActualCount() {
+							t.Fatalf("Expected number of invocations %d different than actual value %d",
+								tc.expectedTries, mockServer.ActualCount())
+						}
+					})
+				})
+			}
+		})
```

Comment:
I think it's better to have a separate nested `when` for each test case so it'll be easier to investigate failures in the future (the output will indicate the specific use case that failed).
In order to prevent code duplication, you can create some helper functions.

Reply:
I refactored the code and put each test case on a separate `when`,  `it` statement. I didn't find a good way to prevent code duplication without affecting the readability, I kept each test case with its own code.